File: theory/physics/contemporary_physics/quant_comp_note/04_project/variational_param_evolution/variational_optimization/variational_optimization_03/src/get_op.py
from qiskit.quantum_info import SparsePauliOp


QUBITNUM = 8


def _xx(i):
    return SparsePauliOp.from_sparse_list(
        [
            ("XX", [i, i + 1], 1.0),
        ],
        num_qubits=QUBITNUM,
    )


def _yy(i):
    return SparsePauliOp.from_sparse_list(
        [
            ("YY", [i, i + 1], 1.0),
        ],
        num_qubits=QUBITNUM,
    )


def _zz(i):
    return SparsePauliOp.from_sparse_list(
        [
            ("ZZ", [i, i + 1], 1.0),
        ],
        num_qubits=QUBITNUM,
    )


def _link(i):
    # Δ = 1 (Heisenberg)
    return _xx(i) + _yy(i) + _zz(i)


def get_xyz():
    # Δ = 1 (Heisenberg), 区分奇偶键: 返回 (H_odd, H_even)
    pair_num = QUBITNUM // 2
    Ho = 0.0 * _xx(0)
    He = 0.0 * _xx(0)
    for pair_idx in range(pair_num - 1):
        i = 2 * pair_idx
        j = 2 * pair_idx + 1
        Ho += _link(i)
        He += _link(j)
    Ho += _link(2 * pair_num - 2)

    return Ho.simplify(), He.simplify()


def get_ssh_H(s):
    Ho, He = get_xyz()
    ssh_op = Ho + s * He

    return ssh_op.simplify()


def get_Ui(op):
    if op == "X":
        Ui = SparsePauliOp.from_sparse_list(
            [
                ("X" * QUBITNUM, [i for i in range(QUBITNUM)], 1.0),
            ],
            num_qubits=QUBITNUM,
        )
    elif op == "Z":
        Ui = SparsePauliOp.from_sparse_list(
            [
                ("Z" * QUBITNUM, [i for i in range(QUBITNUM)], 1.0),
            ],
            num_qubits=QUBITNUM,
        )
    else:
        raise ValueError("The op must be X or Z")
    return Ui


def get_ssh_constrained_H(s, ϵ=1):
    H_c = get_ssh_H(s)
    H_c -= ϵ * (get_Ui("X") + 2 * get_Ui("Z"))
    return H_c.simplify()

File: theory/physics/contemporary_physics/quant_comp_note/04_project/variational_param_evolution/variational_optimization/variational_optimization_03/src/get_op.jl
using QuantumToolbox
function get_xyz(qubit_num::Int)
    @assert iszero(rem(qubit_num, 4)) && !iszero(qubit_num) "The number of qbits must in 4N^+."
    # Params settings
    pair_num = qubit_num ÷ 2
    # Qobj preparation
    X = sigmax()
    Y = sigmay()
    Z = sigmaz()
    qubit_dims = ntuple(_ -> 2, qubit_num)
    Ho = 0 * qeye(2^qubit_num; dims=qubit_dims)
    He = 0 * qeye(2^qubit_num; dims=qubit_dims)
    # Get ssh op
    XX(i) = multisite_operator(
        Val(qubit_num),
        i => X, i + 1 => X,
    )
    YY(i) = multisite_operator(
        Val(qubit_num),
        i => Y, i + 1 => Y,
    )
    ZZ(i) = multisite_operator(
        Val(qubit_num),
        i => Z, i + 1 => Z,
    )
    # Δ = 1 (Heisenberg), 区分奇偶键: 返回 (H_odd, H_even)
    link(i) = XX(i) + YY(i) + ZZ(i)
    for pair_idx = 1:(pair_num-1)
        i = 2pair_idx - 1
        j = 2pair_idx
        Ho += link(i)
        He += link(j)
    end
    Ho += link(2pair_num - 1)

    return Ho, He
end


function get_ssh_H(qubit_num::Int, s::Real)
    Ho, He = get_xyz(qubit_num)
    ssh_op = Ho + s * He

    return ssh_op
end

function get_ssh_constrained_H(qubit_num::Int, s::Real; ϵ=1)
    H_c = get_ssh_H(
        qubit_num, s
    )
    H_c -= ϵ * (get_Ui(qubit_num, sigmax()) + 2 * get_Ui(qubit_num, sigmaz()))
    return H_c
end

function get_Ui(qubit_num::Int, op::Qobj)
    Ui = multisite_operator(
        Val(qubit_num),
        (i => op for i in 1:qubit_num)...
    )
    return Ui
end



